File: python/mirage/commands/cli/builtin/gh/template.py
import json
import re
from typing import Any

from mirage.commands.cli.builtin.gh.constants import (TEMPLATE_ACTION,
                                                      TEMPLATE_DECLARATION,
                                                      TEMPLATE_TOKEN)

Variables = dict[str, list[Any]]


def _text(value: Any) -> str:
    if value is None:
        return "<no value>"
    if isinstance(value, bool):
        return str(value).lower()
    if isinstance(value, list):
        return "[" + " ".join(_text(item) for item in value) + "]"
    if isinstance(value, dict):
        return "map[" + " ".join(f"{k}:{_text(value[k])}"
                                 for k in sorted(value)) + "]"
    return str(value)


def _value(token: str, dot: Any, root: Any, variables: Variables) -> Any:
    if token.startswith('"'):
        return json.loads(token)
    if token.startswith('`'):
        return token[1:-1]
    if token in ("true", "false"):
        return token == "true"
    if re.fullmatch(r'-?\d+', token):
        return int(token)
    if token == ".":
        return dot
    path = token.lstrip(".")
    value = dot
    if token.startswith("$"):
        name, _, path = token.partition(".")
        if name != "$" and name not in variables:
            raise ValueError(f'template: undefined variable "{name}"')
        value = root if name == "$" else variables[name][0]
    for key in path.split("."):
        if not key:
            continue
        value = value.get(key) if isinstance(value, dict) else None
    return value


def _call(name: str, args: list[Any]) -> Any:
    if name == "len":
        return len(args[0])
    if name == "join":
        return str(args[0]).join(_text(x) for x in args[1])
    if name == "pluck":
        return [row.get(args[0]) for row in args[1]]
    if name in ("print", "println"):
        return (" " if name == "println" else "").join(
            _text(x) for x in args) + ("\n" if name == "println" else "")
    if name == "printf":
        values = iter(args[1:])

        def format_value(match: re.Match[str]) -> str:
            flag = match[1]
            if flag == "%":
                return "%"
            value = next(values)
            if flag == "q":
                return json.dumps(value, ensure_ascii=False)
            if flag == "f":
                return f"{float(value):f}"
            return _text(value)

        return re.sub(r"%(%|s|v|d|q|f)", format_value, str(args[0]))
    if name == "index":
        value = args[0]
        for key in args[1:]:
            value = value[key]
        return value
    if name == "eq":
        return any(args[0] == value for value in args[1:])
    if name == "ne":
        return args[0] != args[1]
    if name == "not":
        return not args[0]
    if name in ("and", "or"):
        return all(args) if name == "and" else any(args)
    if name in ("color", "autocolor", "hyperlink"):
        return args[-1]
    if name == "truncate":
        n, text = int(args[0]), _text(args[1])
        return text if len(text) <= n else text[:max(0, n -
                                                     3)] + "." * min(n, 3)
    if name == "contains":
        return args[0] in args[1]
    if name == "hasPrefix":
        return str(args[1]).startswith(args[0])
    if name == "hasSuffix":
        return str(args[1]).endswith(args[0])
    raise ValueError(f'template: function "{name}" not defined')


def _eval(expression: str, dot: Any, root: Any, variables: Variables) -> Any:
    tokens = TEMPLATE_TOKEN.findall(expression)
    parts: list[list[str]] = [[]]
    for token in tokens:
        if token == "|":
            parts.append([])
        else:
            parts[-1].append(token)
    value = None
    for i, part in enumerate(parts):
        if not part:
            raise ValueError("template: empty pipeline")
        args = [_value(token, dot, root, variables) for token in part[1:]]
        if i:
            args.append(value)
        value = _value(
            part[0], dot, root, variables) if len(part) == 1 and not i and (
                part[0].startswith(
                    ('.', '$', '"', '`')) or part[0] in ('true', 'false')
                or re.fullmatch(r'-?\d+', part[0])) else _call(part[0], args)
    return value


def _declaration(expression: str) -> tuple[list[str], str, bool]:
    match = TEMPLATE_DECLARATION.fullmatch(expression)
    if match is None:
        return [], expression, False
    return [name for name in match.group(1, 2)
            if name], match[4], match[3] == "="


def render_template(template: str, value: Any) -> str:
    """Render structured output with Go-style actions, pipelines and blocks.

    Args:
        template (str): output template.
        value (Any): selected JSON fields.
    """
    tokens: list[tuple[str, str]] = []
    end = 0
    trim = False
    for match in TEMPLATE_ACTION.finditer(template):
        text = template[end:match.start()]
        if trim:
            text = text.lstrip()
        if match[1]:
            text = text.rstrip()
        tokens.append(("text", text))
        tokens.append(("action", match[2]))
        trim = bool(match[3])
        end = match.end()
    tokens.append(
        ("text", template[end:].lstrip() if trim else template[end:]))

    def render(start: int, stop: int, dot: Any, variables: Variables) -> str:
        output = []
        i = start
        while i < stop:
            tag, action = tokens[i]
            i += 1
            if tag == "text":
                output.append(action)
                continue
            command, _, expression = action.partition(" ")
            if command in ("range", "if", "with"):
                depth, cursor, alternate = 1, i, None
                while cursor < stop:
                    if tokens[cursor][0] == "action":
                        nested = tokens[cursor][1].split(" ", 1)[0]
                        if nested in ("range", "if", "with"):
                            depth += 1
                        elif nested == "end":
                            depth -= 1
                            if depth == 0:
                                break
                        elif nested == "else" and depth == 1:
                            alternate = cursor
                    cursor += 1
                if depth:
                    raise ValueError("template: unexpected EOF")
                names, pipeline, _ = _declaration(expression)
                if len(names) > 1 and command != "range":
                    raise ValueError(
                        f"template: too many declarations in {command}")
                resolved = _eval(pipeline, dot, value, variables)
                body_end = alternate if alternate is not None else cursor
                scope = dict(variables)
                if names and command != "range":
                    scope[names[0]] = [resolved]
                if command == "range" and resolved:
                    entries = [
                        (k, resolved[k]) for k in sorted(resolved)
                    ] if isinstance(resolved, dict) else enumerate(resolved)
                    for key, item in entries:
                        bound = dict(variables)
                        if len(names) == 2:
                            bound[names[0]] = [key]
                        if names:
                            bound[names[-1]] = [item]
                        output.append(render(i, body_end, item, bound))
                elif resolved:
                    output.append(
                        render(i, body_end,
                               resolved if command == "with" else dot, scope))
                elif alternate is not None:
                    output.append(render(alternate + 1, cursor, dot, scope))
                i = cursor + 1
            elif command in ("end", "else"):
                raise ValueError(f"template: unexpected {command}")
            else:
                names, pipeline, assign = _declaration(action)
                result = _eval(pipeline, dot, value, variables)
                if not names:
                    output.append(_text(result))
                elif not assign:
                    variables[names[0]] = [result]
                elif names[0] in variables:
                    variables[names[0]][0] = result
                else:
                    raise ValueError(
                        f'template: undefined variable "{names[0]}"')
        return "".join(output)

    return render(0, len(tokens), value, {})

File: typescript/packages/core/src/commands/cli/builtin/gh/template.ts
import { compareCodePoints } from '../../../../utils/sort.ts'
import type { JsonValue as Value } from '../../../../types.ts'
function text(value: Value | undefined): string {
  if (value == null) return '<no value>'
  if (Array.isArray(value)) return `[${value.map(text).join(' ')}]`
  if (typeof value === 'object')
    return `map[${Object.keys(value)
      .sort(compareCodePoints)
      .map((key) => `${key}:${text(value[key])}`)
      .join(' ')}]`
  return String(value)
}
function lookup(token: string, dot: Value, root: Value): Value {
  if (token.startsWith('"')) return JSON.parse(token) as Value
  if (token.startsWith('`')) return token.slice(1, -1)
  if (token === 'true' || token === 'false') return token === 'true'
  if (/^-?\d+$/.test(token)) return Number(token)
  if (token === '.') return dot
  let value = token.startsWith('$') ? root : dot
  for (const key of token
    .replace(/^\$?\.?/, '')
    .split('.')
    .filter(Boolean))
    value =
      typeof value === 'object' && value !== null && !Array.isArray(value)
        ? (value[key] ?? null)
        : null
  return value
}
function call(name: string, args: Value[]): Value {
  const a = args[0],
    b = args[1]
  switch (name) {
    case 'len':
      return typeof a === 'string' || Array.isArray(a)
        ? a.length
        : Object.keys(a as Record<string, Value>).length
    case 'join':
      return (b as Value[]).map(text).join(text(a))
    case 'pluck':
      return (b as Record<string, Value>[]).map((row) => row[text(a)] ?? null)
    case 'print':
      return args.map(text).join('')
    case 'println':
      return `${args.map(text).join(' ')}\n`
    case 'printf': {
      let i = 1
      return text(a).replace(/%(%|s|v|d|q|f)/g, (_, flag: string) =>
        flag === '%'
          ? '%'
          : flag === 'q'
            ? JSON.stringify(args[i++])
            : flag === 'f'
              ? Number(args[i++]).toFixed(6)
              : text(args[i++]),
      )
    }
    case 'index': {
      let v = a
      for (const key of args.slice(1))
        v = Array.isArray(v) ? v[Number(key)] : (v as Record<string, Value>)[text(key)]
      return v ?? null
    }
    case 'eq':
      return args.slice(1).some((value) => value === a)
    case 'ne':
      return a !== b
    case 'not':
      return !a
    case 'and':
      return args.every(Boolean)
    case 'or':
      return args.some(Boolean)
    case 'color':
    case 'autocolor':
    case 'hyperlink':
      return args.at(-1) ?? ''
    case 'truncate': {
      const n = Number(a),
        s = text(b)
      return s.length <= n ? s : s.slice(0, Math.max(0, n - 3)) + '.'.repeat(Math.min(n, 3))
    }
    case 'contains':
      return text(b).includes(text(a))
    case 'hasPrefix':
      return text(b).startsWith(text(a))
    case 'hasSuffix':
      return text(b).endsWith(text(a))
    default:
      throw new Error(`template: function "${name}" not defined`)
  }
}
function evaluate(expression: string, dot: Value, root: Value): Value {
  const tokens = expression.match(/"(?:\\.|[^"\\])*"|`[^`]*`|[^\s|]+|\|/g) ?? []
  const parts: string[][] = [[]]
  for (const token of tokens) {
    if (token === '|') parts.push([])
    else parts.at(-1)?.push(token)
  }
  let value: Value = null
  parts.forEach((part, i) => {
    const head = part[0]
    if (head === undefined) throw new Error('template: empty pipeline')
    const args = part.slice(1).map((token) => lookup(token, dot, root))
    if (i > 0) args.push(value)
    value =
      part.length === 1 && i === 0 && (/^[.$"`]/.test(head) || /^(true|false|-?\d+)$/.test(head))
        ? lookup(head, dot, root)
        : call(head, args)
  })
  return value
}

/** Render selected JSON with Go-style actions, pipelines and blocks. */
export function renderTemplate(template: string, value: Value): string {
  const tokens: [string, string][] = []
  let end = 0,
    trim = false
  for (const match of template.matchAll(/{{(-?)\s*(.*?)\s*(-?)}}/gs)) {
    let segment = template.slice(end, match.index)
    if (trim) segment = segment.trimStart()
    if (match[1]) segment = segment.trimEnd()
    tokens.push(['text', segment], ['action', match[2] ?? ''])
    trim = Boolean(match[3])
    end = match.index + match[0].length
  }
  tokens.push(['text', trim ? template.slice(end).trimStart() : template.slice(end)])
  function render(start: number, stop: number, dot: Value): string {
    const output: string[] = []
    for (let i = start; i < stop; ) {
      const [tag, action] = tokens[i++] ?? ['', '']
      if (tag === 'text') {
        output.push(action)
        continue
      }
      const command = action.split(' ')[0] ?? '',
        expression = action.slice(command.length + 1)
      if (['range', 'if', 'with'].includes(command)) {
        let depth = 1,
          cursor = i,
          alternate: number | undefined
        for (; cursor < stop; cursor++) {
          const [nestedTag, nestedAction] = tokens[cursor] ?? ['', '']
          if (nestedTag !== 'action') continue
          const nested = nestedAction.split(' ')[0] ?? ''
          if (['range', 'if', 'with'].includes(nested)) depth++
          else if (nested === 'end') {
            if (--depth === 0) break
          } else if (nested === 'else' && depth === 1) alternate = cursor
        }
        if (depth) throw new Error('template: unexpected EOF')
        const resolved = evaluate(expression, dot, value),
          bodyEnd = alternate ?? cursor
        const truthy = Boolean(resolved) && (!Array.isArray(resolved) || resolved.length > 0)
        if (command === 'range' && truthy) {
          const entries = Array.isArray(resolved)
            ? resolved
            : Object.keys(resolved as Record<string, Value>)
                .sort(compareCodePoints)
                .map((key) => (resolved as Record<string, Value>)[key] ?? null)
          output.push(...entries.map((item) => render(i, bodyEnd, item)))
        } else if (truthy) output.push(render(i, bodyEnd, command === 'with' ? resolved : dot))
        else if (alternate !== undefined) output.push(render(alternate + 1, cursor, dot))
        i = cursor + 1
      } else if (command === 'end' || command === 'else')
        throw new Error(`template: unexpected ${command}`)
      else output.push(text(evaluate(action, dot, value)))
    }
    return output.join('')
  }
  return render(0, tokens.length, value)
}
